pingcap/tidb
# store/kv: fix panic for EOF-stream & add recover to copIterateWorker

### What problem does this PR solve? 

#### 1. EOF-stream panic

fixes #8100 's TiDB panic question, that can be reproduce by

- enable stream
- make a smaller `minLogCopTaskTime` in TiDB
- start TiDB-Server and see panic

#### 2. panic make TiDB exit

### What is changed and how it works?

#### 1. EOF-stream panic

https://github.com/pingcap/tidb/blob/32b1dbd8d59fbd725ec3f2971da2323cae0bf885/store/tikv/client.go#L281

```
	// Read the first streaming response to get CopStreamResponse.
	// This can make error handling much easier, because SendReq() retry on
	// region error automatically.
	var first *coprocessor.Response
	first, err = copStream.Recv()
	if err != nil {
		if errors.Cause(err) != io.EOF {
			return nil, errors.Trace(err)
		}
		log.Debug("copstream returns nothing for the request.")
	}
	copStream.Response = first
```

kv in stream mode may return EOF and make response be `nil` and log logic doesn't take about that...

so fix just add nil check in log logic

#### 2. panic make TiDB exit

add recover to copIterateWorker run, alway close channel and send error response if panic occured.

### Check List 

Tests 

 - Manual test (add detailed scripts or steps below)

```
- enable stream
- make a smaller `minLogCopTaskTime` in TiDB
- start TiDB-Server and doesn't saw panic in log and TiDB doesn't exit.
```

Code changes

- add nil-check
- add recover

Side effects

 - add a annoy function call and defer

Related changes

 - Need to cherry-pick to the release branch


---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/pingcap/tidb/8297)
